File: backend/app/schemas/execution.py
"""
Pydantic schemas for Execution model.
"""
from datetime import datetime
from typing import Optional, Dict, Any, List
from uuid import UUID
from pydantic import BaseModel

from app.models.execution import ExecutionStatus


class ExecutionBase(BaseModel):
    """Base execution schema with common fields."""
    symbol: Optional[str] = None


class ExecutionCreate(ExecutionBase):
    """Schema for creating a new execution."""
    pipeline_id: UUID
    mode: Optional[str] = "paper"  # "live", "paper", "simulation", "validation"


class ExecutionUpdate(BaseModel):
    """Schema for updating an execution."""
    status: Optional[ExecutionStatus] = None
    result: Optional[Dict[str, Any]] = None
    error_message: Optional[str] = None
    cost: Optional[float] = None


class ExecutionInDB(ExecutionBase):
    """Schema for execution in database."""
    id: UUID
    pipeline_id: UUID
    user_id: UUID
    status: ExecutionStatus
    mode: str
    result: Dict[str, Any]
    error_message: Optional[str] = None
    cost: float
    logs: Optional[List[Dict[str, Any]]] = None
    agent_states: Optional[List[Dict[str, Any]]] = None
    reports: Optional[Dict[str, Any]] = None
    cost_breakdown: Optional[Dict[str, Any]] = None
    started_at: Optional[datetime] = None
    completed_at: Optional[datetime] = None
    created_at: datetime

    class Config:
        from_attributes = True


class Execution(ExecutionInDB):
    """Schema for execution response."""
    pass


class ExecutionList(BaseModel):
    """Schema for list of executions."""
    executions: List[Execution]
    total: int


class ExecutionSummary(BaseModel):
    """Schema for execution summary (list view)."""
    id: UUID
    pipeline_id: UUID
    pipeline_name: str
    status: ExecutionStatus
    mode: str
    symbol: Optional[str] = None
    trigger_mode: Optional[str] = None  # "signal" or "periodic"
    scanner_name: Optional[str] = None  # Scanner name for signal-based pipelines
    started_at: Optional[datetime] = None
    completed_at: Optional[datetime] = None
    duration_seconds: Optional[float] = None
    total_cost: float
    agent_count: int
    agents_completed: int
    error_message: Optional[str] = None


class ExecutionStats(BaseModel):
    """Schema for execution statistics."""
    total_executions: int
    running_executions: int
    completed_executions: int
    failed_executions: int
    total_cost: float
    avg_duration_seconds: float
    success_rate: float

File: KuberTrading/KuberTrading/Models/Domain/Execution.swift
import Foundation

// MARK: - Full Execution (detail endpoint)

struct Execution: Codable, Identifiable {
    let id: String
    let pipelineId: String
    let pipelineName: String?
    let userId: String?
    var status: String
    let mode: String
    let symbol: String?
    let triggerMode: String?
    let scannerName: String?
    let startedAt: String?
    var completedAt: String?
    var errorMessage: String?
    var result: AnyCodable?
    var costBreakdown: CostBreakdown?
    var agentStates: [AgentState]?
    var logs: [ExecutionLog]?
    var reports: [String: AgentReport]?
    var executionArtifacts: AnyCodable?
    var approvalStatus: String?
    var approvalRequestedAt: String?
    var approvalExpiresAt: String?
    let createdAt: String?
    let updatedAt: String?
    let cost: Double?
    // Detail-only fields
    let executionPhase: String?
    let nextCheckAt: String?
    let monitorIntervalMinutes: Int?
    let pipelineConfig: AnyCodable?
}

// MARK: - Execution Summary (list endpoint)

struct ExecutionSummary: Codable, Identifiable {
    let id: String
    let pipelineId: String
    let pipelineName: String?
    let status: String
    let mode: String
    let symbol: String?
    let triggerMode: String?
    let scannerName: String?
    let startedAt: String?
    let completedAt: String?
    let durationSeconds: Double?
    let totalCost: Double?
    let agentCount: Int?
    let agentsCompleted: Int?
    let errorMessage: String?
    let strategyAction: String?
    let strategyConfidence: Double?
    let tradeOutcome: String?
    let result: AnyCodable?
    let reports: AnyCodable?
}

// MARK: - Cost Breakdown

struct CostBreakdown: Codable {
    let totalCost: Double?
    let llmCost: Double?
    let agentRentalCost: Double?
    let apiCallCost: Double?
    let byAgent: [String: Double]?
}

// MARK: - Agent State

struct AgentState: Codable, Identifiable {
    var id: String { agentId ?? UUID().uuidString }
    let agentId: String?
    let agentType: String?
    let agentName: String?
    var status: String?
    let startedAt: String?
    let completedAt: String?
    let errorMessage: String?
    let output: AnyCodable?
    let cost: Double?
}

// MARK: - Execution Log

struct ExecutionLog: Codable, Identifiable {
    var id: String { "\(executionId ?? "")-\(timestamp ?? "")-\(message.prefix(20))" }
    let executionId: String?
    let timestamp: String?
    let level: String?
    let agentType: String?
    let message: String
    let details: AnyCodable?
}

// MARK: - Agent Report

struct AgentReport: Codable {
    let agentId: String?
    let agentType: String?
    let title: String?
    let summary: String?
    let status: String?
    let details: String?
    let metrics: [AgentReportMetric]?
    let data: AnyCodable?
    let createdAt: String?
}

struct AgentReportMetric: Codable {
    let label: String
    let value: AnyCodable
    let type: String?
}

// MARK: - Execution Stats

struct ExecutionStats: Codable {
    let totalExecutions: Int?
    let runningExecutions: Int?
    let completedExecutions: Int?
    let failedExecutions: Int?
    let totalCost: Double?
    let avgDurationSeconds: Double?
    let successRate: Double?
}

// MARK: - Execution Create Request

struct ExecutionCreate: Codable {
    let pipelineId: String
    let mode: String?
}
